src/main/java/net/neoforged/neoforge/common/IOUtilities.java

```diff
@@ -0,0 +1,160 @@
+     * @param path The path to write the NBT to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void writeNbtCompressed(CompoundTag tag, Path path) throws IOException {
+        atomicWrite(path, stream -> {
+            try (var bufferedStream = new BufferedOutputStream(stream)) {
+                NbtIo.writeCompressed(tag, bufferedStream);
+            }
+        });
+    }
+
+    /**
+     * Behaves much the same as {@link NbtIo#write(CompoundTag, Path)},
+     * but uses {@link #atomicWrite(Path, WriteCallback)} behind the scenes to
+     * ensure the data is stored resilliently.
+     *
+     * @param tag  The tag to write.
+     * @param path The path to write the NBT to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void writeNbt(CompoundTag tag, Path path) throws IOException {
+        atomicWrite(path, stream -> {
+            try (var bufferedStream = new BufferedOutputStream(stream);
+                    var dataStream = new DataOutputStream(bufferedStream)) {
+                NbtIo.write(tag, dataStream);
+            }
+        });
+    }
+
+    /**
+     * Writes data to the given path "atomically", such that a crash will not
+     * leave the the file containing corrupted or otherwise half-written data.
+     * <p>
+     * This method operates by creating a temporary file, writing to that file,
+     * and then moving the temporary file to the correct location after flushing.
+     * If a crash occurs during this process, the temporary file will be
+     * abandoned.
+     * <p>
+     * Furthermore, the stream passed to {@code writeCallback} is not buffered,
+     * and it is the handler's responsibility to implement any buffering on top
+     * of this method.
+     *
+     * @param targetPath    The desired path to write to.
+     * @param writeCallback A callback which receives the opened stream to write
+     *                      data to.
+     *
+     * @throws IOException if an I/O error occurs during writing.
+     */
+    public static void atomicWrite(Path targetPath, WriteCallback writeCallback) throws IOException {
+        final var tempPath = Files.createTempFile(
+                targetPath.getParent(),
+                targetPath.getFileName().toString(),
+                TEMP_FILE_SUFFIX);
+
+        try {
+            writeImpl(tempPath, targetPath, writeCallback);
+        } catch (Exception first) {
+            // If an exception occurs, we want to try and clean up if we can before rethrowing.
+            try {
+                Files.deleteIfExists(tempPath);
+            } catch (Exception second) {
+                // But if we can't, we suppress the exception.
+                first.addSuppressed(second);
+            }
+
+            throw first;
+        }
+    }
+
+    private static void writeImpl(Path tempFile, Path destination, WriteCallback callback) throws IOException {
```

Comment:
Why another function?

Reply:
Personally, I dislike nested try/catch statements; I removed it though since it's technically unnecessary.